ocaml/parsing/parser.mly

```diff
@@ -3085,15 +3091,27 @@ let_binding_body_no_punning:
          version, even though we are creating a slightly different [core_type].
       *)
       { let exp, poly =
-          wrap_type_annotation ~loc:$sloc ~typloc:$loc($6) $4 $6 $8
+          wrap_type_annotation ~loc:$sloc ~typloc:$loc($6) $4 $6 $9
         in
         let loc = ($startpos($1), $endpos($6)) in
-        (ghpat ~loc (Ppat_constraint($1, poly)), exp, None, [])
+        let modes_ghost = Mode.ghostify modes in
+        let exp = mkexp_with_modes ~ghost:true ~loc:$sloc modes_ghost exp in
+        (ghpat ~loc (Ppat_constraint($1, poly)), exp, None, let_binding_mode_attrs modes)
        }
   | pattern_no_exn EQUAL seq_expr
       { ($1, $3, None, []) }
-  | simple_pattern_not_ident COLON core_type EQUAL seq_expr
-      { ($1, $5, Some(Pvc_constraint { locally_abstract_univars=[]; typ=$3 }), []) }
+  | simple_pattern_not_ident modes = at_mode_expr_nonempty EQUAL seq_expr
+      {
+        let modes_ghost = Mode.ghostify modes in
+        let exp = mkexp_with_modes ~ghost:true ~loc:$sloc modes_ghost $4 in
+        ($1, exp, None, let_binding_mode_attrs modes)
+      }
+  | simple_pattern_not_ident COLON core_type modes=atat_mode_expr EQUAL seq_expr
```

Comment:
Did you consider making a single production for "either `at_mode_expr` OR `COLON core_type atat_mode_expr`"? That would allow you to merge these two rules, potentially.

Reply:
Usually I would try to factor out this, but is it still worth it when it's used only once? (in contrast to `constraint_` which is usually multiple times)

Reply:
I think you should do whatever leads to simpler-feeling code. That might mean factoring out the action into a helper function, factoring out the two rules into a production that can share the same action, or leaving things as-is.